### PR TITLE
Handle related products in new product page

### LINKED ISSUE #25831: Product page v2: handle duplicate attachments
#### Describe the bug

In the new product page, when selecting attachment files you can select the same file multiple times. On submit no error is displayed, and only one attachment is associated but it results in a useless long list and can be confusing as you don't understand only will be added after submit.

#### Expected behavior

Two possible solutions:
- prevent the search component from displaying already selected attachments
- display a growl error when selecting an attachment that is already in the list

#### Steps to Reproduce

1. Go to Produc page v2
2. In the Specifications tab search for an attachment
3. Select one attachment, then perform the same research and attach the same file
4. The attachment is added twice in the list

**Screenshots**

image

#### Additional information

* PrestaShop version: 8.0
* PHP version: N/A

### PR DESCRIPTION
**Pull request by jolelievre**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | The related products are now editable in the new product page They rely on the `EntitySearchInput` component so the PR mainly needed to handle:<br>- a custom template to display products<br>- update provider and an associated command builder<br>- add the missing image in the query for related products (and update behat tests to check all those fields more accurately)<br><br>This PR also include a new feature in the entity search component, it can now filter out the selected items from the remote response, this avoids trying to add the same item twice, this feature is togglable and configurable You can specify which field of the item represents its identity, for now it relies only on a single field which is enough but it might need improvement in the future.
| Type?             | improvement
| Category?         | BO
| BC breaks?        | yes
| Deprecations?     | no
| Fixed ticket?     | ~
| How to test?      | In the new product page try and select some products to associate them as related products, you can also remove this association
| Possible impacts? | ~

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/25920)

### Specs

**Add a related product**: 
The user can search by product name or product reference. You can click on it to select it. An associated product can only once be associated with the same product. You can't associate a product to itself.
There is no limit to the number of associated products. 

When the user starts typing, if there are results, they are displayed and highlighted in the dropdown when the user hovers the results using the keyboard or the mouse.

**Delete a related product**: 
If you want to delete only one associated product, you can click on the delete icon. 

Add and delete actions are only saved when the users save the form.

### Design

Capture d’écran 2021-11-04 à 20 00 34

Once the product added, It displays the product name with its reference and the product cover.

### Breaking Change

The `RelatedProduct` query result class now needs an additional image parameter

### CODE REVIEW

File: admin-dev/themes/new-theme/scss/pages/product/product_page.scss [hunk 1]
```diff
@@ -1075,6 +1075,40 @@ $product-page-padding-bottom: 80px !default;
...
+      display: flex;
+      flex-direction: row;
+      flex-wrap: wrap;
+      margin: 0.5rem -0.5rem;
+    }
+
+    .related-product {
+      display: flex;
+      flex-direction: column;
+      width: 180px;
```
Comment (NeOMakinG) [inline admin-dev/themes/new-theme/scss/pages/product/product_page.scss]: Looks like a there is a problem there, we almost never use a fixed width anywhere because on responsive  it will be too small

Usually it's

```
width: 100%;
max-width: 180px;
```

File: admin-dev/themes/new-theme/scss/pages/product/product_page.scss [hunk 2]
```diff
@@ -1075,6 +1075,40 @@ $product-page-padding-bottom: 80px !default;
...
+
+      &-image {
+        width: 100%;
+
+        img {
+          display: block;
+          width: auto;
+          max-width: 170px;
+          height: auto;
+          max-height: 170px;
```
Comment (NeOMakinG) [inline admin-dev/themes/new-theme/scss/pages/product/product_page.scss]: Depending of the image ratio you choose this may cause some UI issues, better to use the `.img-fluid` class or use the trick I sent before (which is the same as img-fluid) and don't set the height I think so

**Review by NeOMakinG — COMMENTED** [review]
JS looks fine to me, but some SCSS comments

File: admin-dev/themes/new-theme/scss/pages/product/product_page.scss (hunk 2 above)
Comment (jolelievre) [reply admin-dev/themes/new-theme/scss/pages/product/product_page.scss]: Ok that makes sense, thanks!! I'd prefer to use `img-fluid` then but I'm not sure exactly how. I mean I add the `img-fluid` to the image itself, but then where do I limit the width of my blocks?

Do I keep the `width: 180px;` on `.related-product` ? Or as you say use `max-width: 180px;`? Should I have a min-wdth as well? And finally, how do I handled the "margin" on the image, I set 170px so that there is a gap, should I rely on a padding on `.related-product` maybe?

**Review by Progi1984 — DISMISSED** [review]

**Review by NeOMakinG — DISMISSED** [review]
Front looks fine to me

**Review by Progi1984 — DISMISSED** [review]

**Review by NeOMakinG — DISMISSED** [review]

**Review by Progi1984 — DISMISSED** [review]

**Review by NeOMakinG — DISMISSED** [review]

**Review by Progi1984 — DISMISSED** [review]

**Comment by khouloudbelguith** [comment]
Hi @jolelievre,

I have some issues:

**First issue**

The trash button is to delete all related products is not displayed
I found in this [spec](https://github.com/PrestaShop/prestashop-specs/blob/master/content/1.7/back-office/catalog/products/add-edit.md#basic-settings-tab-epic)
`The trash button is to delete all related products. When you click on it, you have a modal for cancel or confirm the action.`

**Second issue**

1. Go to BO > Catalog > Edit a product V2
2. In the `Related products` section
3. Start typing something in the product field
4. Clear the search product field: you can see in the Console tab an error 400 

image
In the network tab, we have this error
image
I attached a screen record https://drive.google.com/file/d/15S8I7PXPLZ49BLax-EJgvyGYK7qRNput/view?usp=sharing


**Third issue**

You can add the product itself as a related product
image

**Fourth issue**

1. Go to BO > Catalog > Edit product V2
2. Add a related product Product1
3. Add a related product Product2
4. Delete the Product1 from the related product section
5. At this step, you can add the same related product Product2 a second time 

image

**Fifth issue**
1. Create a product with a long title and long reference
2. Edit another product V2
3. Try to add the product with the long name in the "Related product section"
4. See error 

image

Thanks!

**Comment by NeOMakinG** [comment]
@jolelievre This is what I told you about the name while on meet, I would add a `text-overflow: ellipsis;` on the description element

**Comment by jolelievre** [comment]
> @jolelievre This is what I told you about the name while on meet, I would add a `text-overflow: ellipsis;` on the description element

Yep I though about it too when I saw the screenshot 😅 I will add the ellipsis!

File: src/PrestaShopBundle/Form/Admin/Sell/Product/SEO/RedirectOptionType.php [hunk 3]
```diff
@@ -155,17 +170,21 @@ public function buildForm(FormBuilderInterface $builder, array $options)
...
-            ],
-            'alert_message' => $this->getRedirectionAlertMessages(),
-        ]);
+        $resolver
+            ->setDefaults([
+                'product_id' => null,
+                'required' => false,
+                'label' => $this->trans('Redirection page', 'Admin.Catalog.Feature'),
+                'label_tag_name' => 'h2',
+                'label_help_box' => $this->trans('When your product is disabled, choose to which page you’d like to redirect the customers visiting its page by typing the product or category name.', 'Admin.Catalog.Help'),
```
Comment (Julievrz) [inline src/PrestaShopBundle/Form/Admin/Sell/Product/SEO/RedirectOptionType.php:179]: I think we can remove this tooltip, it's redundant. :-) The label "Redirection page" and the help text "Redirection when offline" are already clear enough in my opinion.

Comment (jolelievre) [reply src/PrestaShopBundle/Form/Admin/Sell/Product/SEO/RedirectOptionType.php:179]: Hi, I'm not sure this feature is very technical Some merchant don't even know what a "redirection" means And this page already has a lot of different fields to setup, in this case a little extra info doesn't hurt IMHO Especially since it's in a tooltip so it doesn't bother the use of the page unless you put your mouse cursor on it.

See video https://watch.screencastify.com/v/EWaJZ5bf8cCOwWXz2G6s

Comment (jolelievre) [reply src/PrestaShopBundle/Form/Admin/Sell/Product/SEO/RedirectOptionType.php:179]: As discussed with @Julievrz This part can be improved (too many blocks and/or tooltip on this page). But this improvement will be done later globally on the whole page For now this PR can be merged as-is to avoid blocking coming development regarding pakcs

**Comment by Julievrz** [comment]
Wording ✅ for now but as discussed with @jolelievre improvements should be added later on.

**Review by NeOMakinG — DISMISSED** [review]

File: admin-dev/themes/new-theme/js/components/entity-search-input.ts [hunk 4]
```diff
@@ -418,4 +447,20 @@ export default class EntitySearchInput {
 
     return template;
   }
+
+  private getSelectedIds(): any[] {
```
Comment (NeOMakinG) [inline admin-dev/themes/new-theme/js/components/entity-search-input.ts]: An input value is always a string, I think you can replace this and types above with `Array<string>` or `string[]`

File: admin-dev/themes/new-theme/js/components/entity-search-input.ts [hunk 5]
```diff
@@ -303,26 +309,49 @@ export default class EntitySearchInput {
...
         cache: false,
         wildcard: this.options.queryWildcard,
-        transform(response: any) {
+        transform: (response: any) => {
           if (!response) {
             return [];
           }
-          return response;
+
+          const selectedIds: any[] = this.getSelectedIds();
```
Comment (NeOMakinG) [inline admin-dev/themes/new-theme/js/components/entity-search-input.ts]: Same thing here about strings

File: admin-dev/themes/new-theme/js/components/entity-search-input.ts (hunk 4 above)
Comment (jolelievre) [reply admin-dev/themes/new-theme/js/components/entity-search-input.ts]: Even for a `<input type="number" />` ?

Comment (NeOMakinG) [reply admin-dev/themes/new-theme/js/components/entity-search-input.ts]: Yep, it's a string :/

image

Comment (NeOMakinG) [reply admin-dev/themes/new-theme/js/components/entity-search-input.ts]: JavaScript 🤝

**Review by Progi1984 — DISMISSED** [review]

**Review by NeOMakinG — DISMISSED** [review]

**Review by Julievrz — DISMISSED** [review]

**Comment by khouloudbelguith** [comment]
Hi @jolelievre,

The **First issue** still exists

> The trash button is to delete all related products is not displayed
I found in this [spec](https://github.com/PrestaShop/prestashop-specs/blob/master/content/1.7/back-office/catalog/products/add-edit.md#basic-settings-tab-epic)
`The trash button is to delete all related products. When you click on it, you have a modal for cancel or confirm the action.`

This is an issue or we should update the spec?

I found some other issues:

**Second issue**
1. Go to BO > Catalog > Edit a product V2
2. In the `Related products` section
3. Search not-existing product
4. you can see in the Console tab an error 400 
image

**Third issue**
1. Create a lot of products with the same name (I imported a file csv) for example `test`
image
2. Go to BO > Catalog > Edit a product V2
2. In the `Related products` section
3. Search this product
4. See error: only 3 products displayed 
image

On the product page V1, we have a scroll
image

Thanks!

**Comment by Julievrz** [comment]
As discussed with @MatShir, I'm putting back the "waiting for wording" label to keep track of the wordings that need to be modified later on. But this is not blocking the PR.

**Comment by jolelievre** [comment]
Hi @khouloudbelguith 

Thanks for the feedback.

#### First issue

I discussed this with @MatShir and indeed we decided that the remove all button was not really useful, so the specifications should be updated.

#### Second issue

This is on purpose, the search API follow the recent convention used in REST API to rely on the HTTP status code, so when no items are found instead of returning a request with status code 200 and empty content we return a response with 404 error code which identifies clearly the "error" Even if it is displayed as an error in the javascript console (default behaviour) it doesn't break the JS behaviour since it correctly displays that `No results found for "sdfdf"` message

This should be validated by @PrestaShop/prestashop-core-developers I guess.

#### Third issue

I tried to reproduce it but I have the scrollbar in the new page as well, see video:
https://watch.screencastify.com/v/QJnxB3khjhBwGiXNBftE

That's weird... maybe a problem with built assets or cache?

**Comment by khouloudbelguith** [comment]
Hi @jolelievre,

Thanks for the clarifications
> Third issue
I tried to reproduce it but I have the scrollbar in the new page as well, see video:
https://watch.screencastify.com/v/QJnxB3khjhBwGiXNBftE
That's weird... maybe a problem with built assets or cache?

I will try to re-install the PR and recheck.
>Edit

@jolelievre, there are conflicting files, I couldn't make a checkout to this PR. 
image


Thanks to check the Conflicting files!

**Review by Progi1984 — DISMISSED** [review]

**Review by NeOMakinG — DISMISSED** [review]

**Review by Progi1984 — DISMISSED** [review]

**Review by NeOMakinG — DISMISSED** [review]

**Comment by khouloudbelguith** [comment]
Hi @jolelievre,

I have some issues:

**First issue**
1. Enable  Experimental product page
2. Go to BO > Catalog > Products > Edit a product V2
3. In the `Related product` section > try to search not existing product for example `xxxx`
4. See an error in the console 
image

I attached a screen record
https://watch.screencastify.com/v/QGU0TsrbArtoEKOtc9Uw

**Second issue**
>The user can search by product name or product reference. 

1. Enable  Experimental product page
2. Go to BO > Catalog > Products > Edit a product V2
3. In the `Related product` section > try to search with `demo`
4. Only one product displayed 
image
 => in the standard page => we have the list of al products 
image

**Third issue**
1. Import this file csv 
[products_import.csv.txt](https://github.com/PrestaShop/PrestaShop/files/7475844/products_import.csv.txt)
2. Go to BO > Catalog > Products > Edit a product V2
3. In the `Related product` section > try to search by `test`
4. I have no result 
image

I have this list 
image

I have some issues related to the spec:

1. There is no **placeholder**="Search a product" 
image
2. In the spec there is no `search-with-icon`, in our case, there is a search icon 
3. In the spec, the icon is displayed at the `right`, in our case, the search icon at the `left`

**Spec**
image
**With PR**
image

Thanks!

**Comment by TristanLDD** [comment]
Hello! About the UI, everything is ok expect for the search icon that is now on the left of search inputs according to the UI kit: 
Capture d’écran 2021-11-04 à 17 42 53

Thanks!

File: src/PrestaShopBundle/Form/Admin/Sell/Product/SEO/SEOType.php [hunk 6]
```diff
@@ -243,11 +245,15 @@ private function getFriendlyAlterMessages(): array
...
-        $resolver->setDefaults([
-            'label' => $this->trans('Search Engine Optimization', 'Admin.Catalog.Feature'),
-            'label_tag_name' => 'h2',
-            'label_subtitle' => $this->trans('Improve your ranking and how your product page will appear in search engines results.', 'Admin.Catalog.Feature'),
-            'required' => false,
-        ]);
+        $resolver
+            ->setDefaults([
+                'product_id' => null,
+                'label' => $this->trans('Search Engine Optimization', 'Admin.Catalog.Feature'),
```
Comment (Julievrz) [inline src/PrestaShopBundle/Form/Admin/Sell/Product/SEO/SEOType.php]:
```suggestion
                'label' => $this->trans('Search engine optimization', 'Admin.Catalog.Feature'),
```

**Comment by Julievrz** [comment]
I'm adding the mockups here (with the right wording)

image
image
Second screenshot: Note that this is "Search product" (without an "a")

**Comment by jolelievre** [comment]
Thanks, @khouloudbelguith for the feedback, here are the answers for each point:

#### First issue (404 in console)

I believe I already answered this one in a previous comment but this is on purpose, the API finds no result so a 404 error code is adapted in those cases. It doesn't matter if the develop console displays it as an error (it just considers any 404 as an error). The component handles it correctly since it displays the "No results found" message in that case.

I already discussed this behaviour with @PierreRambaud who agrees this is acceptable and even a good practice.

#### Second issue and Third issue (missing results displayed)

Both issues are related to the same problem, it's actually a bug from the `typeahead.js` library that is used by this component. It doesn't handle well when the number of results returned is less than the set limit. Instead of displaying them all, it returns `limit - numberOfResults` So for example with `demo` search the API returns 19 results, one is filtered out since it matches the current product, so 18 products to display, the previous limit was 20 so it only displayed 2 results.

On the legacy page (where the limit is 30) you can actually see the same bug, when you search for `demo` only 11 results are rendered: 30 - 19 products The displayed products should be 19 because it doesn't filter the current product, unlike the new one since you found this bug (in that sense new page is better).

For the issue with your `tests` products, it's even worse because the number of results exceeds the limit, and in this case, it simply displays nothing...

We don't have a perfect fix for this bug for now because we can't update the library (it hasn't been updated for 7 years...) we need to find a solution for this but it's outside of this PR's scope, it requires a dedicated issue and PR.

So for now I settled with the same configuration as the legacy page:
- 30 display limit
- 20 API limit

So it will at least display 10 results, that matches the legacy page so it's acceptable for now I'd say.

#### Spec issues

As Tristan said the current PR matches the most recent modifications of the design. The only thing I couldn't fix is the search icon that is supposed to be on the left. This is handled by the PrestaShop UI kit which is an external library so I can't fix it in the core. I'm waiting for @NeOMakinG opinion regarding this issue to know if I should update something in this PR or if it's related to another issue.

### TLDR;

With the latest changes, the only thing to check is that now the `demo` and `tests` products are partially displayed, not perfect but at least the same (or better) as on the legacy page.

File: src/PrestaShopBundle/Form/Admin/Sell/Product/SEO/SEOType.php (hunk 6 above)
Comment (jolelievre) [reply src/PrestaShopBundle/Form/Admin/Sell/Product/SEO/SEOType.php]: That's not the scope of this PR but I modified it regardless
Although I'm surprised, since SEO is an acronym shouldn't we use capital for each word of the acronym? `SEO = Search Engin Optimization` ?

**Comment by jolelievre** [comment]
As discussed with @NeOMakinG the search icon is defined to be on the right in the PrestaShop UI kit
https://github.com/PrestaShop/prestashop-ui-kit/blob/b8542c10f0e969ff069d659ed490c323181cbc08/scss/_forms.scss#L210

It's been this way for years now... So if you plan on changing the UI kit it's fine but it's not to be handled in this PR

**Review by NeOMakinG — DISMISSED** [review]

**Comment by TristanLDD** [comment]
My bad, I thought the new search bar was already integrated to the UI kit. I'll open another issue for this specific component then.

File: src/PrestaShopBundle/Form/Admin/Sell/Product/Description/DescriptionType.php [hunk 7]
```diff
@@ -110,9 +139,20 @@ public function buildForm(FormBuilderInterface $builder, array $options)
...
+                'entry_options' => [
+                    'block_prefix' => 'related_product',
+                ],
+                'remote_url' => $this->router->generate('admin_products_v2_search_associations', [
+                    'languageCode' => $this->employeeIsoCode,
+                    'query' => '__QUERY__',
+                ]),
+                'min_length' => 3,
+                'filtered_identities' => $productId > 0 ? [$productId] : [],
+                'placeholder' => $this->trans('Search product', 'Admin.Catalog.Feature'),
```
Comment (Julievrz) [inline src/PrestaShopBundle/Form/Admin/Sell/Product/Description/DescriptionType.php]:
```suggestion
                'placeholder' => $this->trans('Search product', 'Admin.Catalog.Help'),
```

**Review by Julievrz — COMMENTED** [review]
That would be great if you could also remove the tooltip about the Redirection page. :-) I discussed the issue with @TristanLDD and he agrees that we can remove it. The final result seems clear to us: 

image

Also, I think the wording "By default, the page redirects to the main category." is missing.

**Review by Progi1984 — DISMISSED** [review]

File: src/PrestaShopBundle/Form/Admin/Sell/Product/Description/DescriptionType.php (hunk 7 above)
Comment (jolelievre) [reply src/PrestaShopBundle/Form/Admin/Sell/Product/Description/DescriptionType.php]: I'll fix the domain here, but the other modifications are not in the scope of this PR It's been stalling for a long time and I don't want to delay it for a wording that's not related This PR is about related products, not SEO tab, they just share a comment component which is why the `SEOType` file was updated a bit
We will do a full review of all the wording on the page in each tabs later anyway

**Review by Progi1984 — DISMISSED** [review]

**Review by NeOMakinG — DISMISSED** [review]

**Comment by khouloudbelguith** [comment]
Hi @jolelievre,

I found a strange issue following these steps:
1. Enable Experimental product page
2. Go to BO > Catalog > Products > Add/Edit a product V2
3. Check the placeholder="Search product" => **ok**
4. Click on the SEO tab and try to change the `Redirection when offline` to `xxx redirection to a category`
image
5. Go back to the Description Tab, the **placeholder** of the Related product is NOK 
image
6. Click on the SEO tab again and try to change the `Redirection when offline` to `xxx redirection to a product`
image
7. Go back to the Description Tab, the **placeholder** of the Related product is NOK  
image

I checked the Standard page => ok 

Thanks!

**Comment by jolelievre** [comment]
Ho! Nice one @khouloudbelguith thanks

Indeed in the SEO tab the placeholder is dynamically updated, but it seems like the selector is too wide because it also updates the other components in the page 😅

I'm on it!

**Comment by jolelievre** [comment]
Hi @khouloudbelguith 

I just fixed the bug, it's a very small modification (see commit https://github.com/PrestaShop/PrestaShop/pull/25920/commits/a5e6e17fdc45c154fa72d306cde9bc18b55ccbc3) so I think you can test the PR again even if it's not approved yet, just to gain a bit of time 😉

**Review by NeOMakinG — APPROVED** [review]

**Review by kpodemski — APPROVED** [review]

**Review by PierreRambaud — APPROVED** [review]

**Comment by khouloudbelguith** [comment]
Hi @jolelievre,

It is ok :heavy_check_mark: 
- Search (no image - with multiple images (cover image) - image changed )
- Search the same product 
- Search the product with reference - ref combination 
- Search invalid product (404) => ok
- add a related product (with a simple product - a product with combinations - pack product - vector product)
- Remove a related product => ok

image


image
In the FO
image

https://watch.screencastify.com/v/Yi6xKTU9pUMqxXAgOS04
Thanks!

**Comment by jolelievre** [comment]
Thank you @khouloudbelguith

**Comment by matthieu-rolland** [comment]
thank you everyone
